components/brave_wallet/browser/eth_gas_utils.cc

```diff
@@ -0,0 +1,138 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/eth_gas_utils.h"
+
+#include <algorithm>
+#include <cmath>
+
+#include "base/check.h"
+#include "base/logging.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+
+namespace brave_wallet {
+
+namespace eth {
+
+// Assumes there are 3 reward percentiles per element
+// The first for low, the second for avg, and the third for high.
+// The following calculations will be made:
+// - suggested_base_fee_per_gas will be the pending
+//   base_fee_per_gas (last element) * 12.5%
+// - avg_priority_fee will be the 0.4 * length's element of the sorted
+//   reward array.
+// - The same applies to low_prirority_fee, but if it was equal to avg
+//   then we walk it back to the next smallest element if possible.
+// - The same applies to high_prirority_fee, but if it was equal to avg
+//   then we walk it forward to the next biggest element if possible.
+bool GetSuggested1559Fees(const std::vector<std::string>& base_fee_per_gas,
+                          const std::vector<double>& gas_used_ratio,
+                          const std::string& oldest_block,
+                          const std::vector<std::vector<std::string>>& reward,
+                          uint256_t* low_priority_fee,
+                          uint256_t* avg_priority_fee,
+                          uint256_t* high_priority_fee,
+                          uint256_t* suggested_base_fee_per_gas) {
+  CHECK(low_priority_fee);
+  CHECK(avg_priority_fee);
+  CHECK(high_priority_fee);
+  CHECK(suggested_base_fee_per_gas);
+
+  // Not enough info to determine values
+  if (base_fee_per_gas.empty())
+    return false;
+
+  // "pending" is the last element in base_fee_per_gas
+  // "latest" is the 2nd last element in base_fee_per_gas
+  std::string pending_base_fee_per_gas = base_fee_per_gas.back();
+  uint256_t pending_base_fee_per_gas_uint;
+  if (!HexValueToUint256(pending_base_fee_per_gas,
+                         &pending_base_fee_per_gas_uint))
+    return false;
+
+  // We use "double" math and this is unlikely to get hit, so return false
+  // if the value is too big.
+  if (pending_base_fee_per_gas_uint > kMaxSafeIntegerUint64)
+    return false;
+
+  // Compiler crashes without these 2 casts :(
+  double pending_base_fee_per_gas_dbl =
+      static_cast<double>(static_cast<uint64_t>(pending_base_fee_per_gas_uint));
+  // The base fee is not part of the RLP. We only specify priority fee and max
+  // fee. So the excess will be refunded and it will be at most 12.5% It's best
+  // to assume a larger value here so there's a better chance it will get in the
+  // next block and if it's too high it will go back to the user
+  pending_base_fee_per_gas_dbl *= 1.125;
```

Comment:
perhaps it would be good to make some kind of an option for a user for that one sometime in the future.

Reply:
The excess will be refunded to the user

Reply:
There could perhaps be a user who wants to send a transaction that doesn't have enough gas so it might not be ideal but it gives a higher chance the transaction will be selected in the next block at no actual cost. I think it is ok to start and we can adjust based on user feedback.  The user can also adjust things manually.

Reply:
yeah, makes sense as a user can always edit gas manually.